### PR TITLE
Cumulus 2123 -- add RDS development cluster to Core

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-2123: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2123)

## Changes

* Updates Core to add an RDS cluster deployment module, and updates data persistence/core example modules to utilize it.    Updates Core to provisionally utilize the database cluster's login credentials and security group via adding the ECS task runner to the database security group and setting the API lambdas to have access to the login credentials and be in the security group.

## PR Checklist

- [ ] Update CHANGELOG
- [x] Unit tests
- [x] Adhoc testing
- [x] Integration tests

### CODE REVIEW

File: lambdas/db-provision-user-database/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      connection: config,
+      acquireConnectionTimeout: 60000,
+    });
+    const dbUser = event.prefix.replace('-', '_');
+
+    const tableExists = await knex.select(1).as('result')
+      .from('pg_database').where('datname', `${dbUser}_db`);
+
+    if (tableExists.length === 0) {
+      await knex.raw(`create user ${dbUser} with encrypted password '${event.dbPassword}'`);
```
Comment (Jkovarik) [inline lambdas/db-provision-user-database/src/index.ts:51]: There are probably some questions here regarding if we want a more complicated schema/permissions structure, however that's probably an additional ticket.   Comments welcome.

File: example/data-persistence-tf/provision-database.tf [hunk 2]
```diff
@@ -0,0 +1,5 @@
+terraform {
```
Comment (markdboyd) [inline example/data-persistence-tf/provision-database.tf:1]: I am probably dense, but what does this file do?

File: example/data-persistence-tf/variables.tf [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  type    = string
+  default = null
+}
+
+variable "prefix" {
+  type = string
+}
+
+variable "rds_access_secret_id" {
+  description = "AWS Secrets Manager secret ID containing jsonified db credentials containing at least host, password, port"
```
Comment (markdboyd) [inline example/data-persistence-tf/variables.tf:13]:
```suggestion
  description = "AWS Secrets Manager secret ID containing a JSON string of DB credentials (at least host, password, port)"
```

File: example/db-migration-tf/variables.tf [hunk 4]
```diff
@@ -7,31 +7,30 @@ variable "permissions_boundary_arn" {
...
-}
-
-variable "pg_host" {
-  type = string
-}
-
-variable "pg_user" {
+variable "rds_access_secret_id" {
   type = string
+  default = null
```
Comment (markdboyd) [inline example/db-migration-tf/variables.tf:12]: What happens if this is null?

File: example/rds-cluster-tf/main.tf [hunk 5]
```diff
@@ -0,0 +1,23 @@
+terraform {
+  required_providers {
+    aws = ">= 2.31.0"
+  }
+}
+
+provider "aws" {
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply example/rds-cluster-tf/main.tf:7]: I think that provider is only there to support deploying a specific a Lambda to us-west-2, which we use to test in-region/out-of-region access

File: example/rds-cluster-tf/main.tf [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  region  = var.region
+  profile = var.aws_profile
+}
+
+module "rds_cluster" {
+  source      = "../../tf-modules/cumulus-rds-tf"
+  profile     = var.profile
+  db_username = var.db_username
+  db_password = var.db_password
+  region      = "us-east-1"
```
Comment (markdboyd) [inline example/rds-cluster-tf/main.tf:17]:
```suggestion
  region      = var.region
```

File: example/rds-cluster-tf/variables.tf [hunk 7]
```diff
@@ -0,0 +1,45 @@
+variable "profile" {
+  description = "AWS profile to use for authentication"
```
Comment (markdboyd) [inline example/rds-cluster-tf/variables.tf:2]: I'm not totally clear on what this means. What is AWS profile referring to?

File: lambdas/db-migration/main.tf [hunk 8]
```diff
@@ -81,7 +84,8 @@ resource "aws_lambda_function" "db_migration" {
     content {
       subnet_ids = var.subnet_ids
       security_group_ids = [
-        aws_security_group.db_migration[0].id
+        aws_security_group.db_migration[0].id,
```
Comment (markdboyd) [inline lambdas/db-migration/main.tf:87]: Not really related to your work but I was wondering why this Lambda is creating its own security group at all?

File: lambdas/db-migration/src/index.ts [hunk 9]
```diff
@@ -16,30 +18,54 @@ const getRequiredEnvVar = (name: string, env: NodeJS.ProcessEnv): string => {
...
 };
 
-const getConnectionConfig = (env: NodeJS.ProcessEnv): Knex.PgConnectionConfig => ({
-  host: getRequiredEnvVar('PG_HOST', env),
-  user: getRequiredEnvVar('PG_USER', env),
-  // TODO Get this value from secrets manager
-  password: getRequiredEnvVar('PG_PASSWORD', env),
-  database: getRequiredEnvVar('PG_DATABASE', env),
-});
+const getSecretConnectionConfig = async (SecretId: string): Promise<Knex.PgConnectionConfig> => {
```
Comment (markdboyd) [inline lambdas/db-migration/src/index.ts:21]: It seems like we're going to need this in several lambdas, so maybe we can put this in the `db` package?

Also, we have this if we want to dogfood our own aws-client package, but I'm not sure if it really improves anything

https://github.com/nasa/cumulus/blob/master/packages/aws-client/src/SecretsManager.ts#L7

File: lambdas/db-migration/src/index.ts [hunk 10]
```diff
@@ -16,30 +18,54 @@ const getRequiredEnvVar = (name: string, env: NodeJS.ProcessEnv): string => {
...
+  const secretsManager = new AWS.SecretsManager();
+  const response = await secretsManager.getSecretValue(
+    { SecretId } as AWS.SecretsManager.GetSecretValueRequest
+  ).promise();
+  if (response.SecretString === undefined) {
+    throw new Error('Database credentials are undefined!');
```
Comment (markdboyd) [inline lambdas/db-migration/src/index.ts:27]: Should we say something more specific about how the secret `SecretId` for database credentials could not be parsed? It would give more helpful information for anyone trying to debug a failure, but I'm not sure if that's leaking too much information. But I think anyone running this Lambda we're okay with seeing that information

File: lambdas/db-migration/src/index.ts [hunk 11]
```diff
@@ -16,30 +18,54 @@ const getRequiredEnvVar = (name: string, env: NodeJS.ProcessEnv): string => {
...
+  const dbAccessMeta = JSON.parse(response.SecretString);
+  return {
+    host: dbAccessMeta.host,
+    user: dbAccessMeta.username,
+    password: dbAccessMeta.password,
+    database: dbAccessMeta.database,
+  };
+};
+
+const getConnectionConfig = async (env: NodeJS.ProcessEnv): Promise<Knex.PgConnectionConfig> => {
```
Comment (markdboyd) [inline lambdas/db-migration/src/index.ts:38]: @marc mentioned that he didn't want to move this helper to a package, but I foresee us using this exact invocation across multiple Lambdas, including data-migration. I could see it moving to the `db` package but it's not essential

File: lambdas/db-migration/src/index.ts [hunk 12]
```diff
@@ -16,30 +18,54 @@ const getRequiredEnvVar = (name: string, env: NodeJS.ProcessEnv): string => {
...
-  });
+    knex = Knex({
+      client: 'pg',
+      connection: connectionConfig,
+      debug: env?.KNEX_DEBUG === 'true',
+      asyncStackTraces: env?.KNEX_ASYNC_STACK_TRACES === 'true',
+      migrations: {
+        directory: path.join(__dirname, 'migrations'),
+      },
+      acquireConnectionTimeout: 60000,
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply lambdas/db-migration/src/index.ts:64]: So handling cold start issues with RDS Serverless? I agree that we can leave it for now and circle back to improve

File: lambdas/db-provision-user-database/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+      SecretString: JSON.stringify({
+        username: dbUser,
+        password: event.dbPassword,
+        engine: 'postgres',
+        database: `${dbUser}_db`,
+        host: config.host,
+        port: config.port,
+      }),
+    }).promise();
+    await knex.destroy();
```
Comment (markdboyd) [inline lambdas/db-provision-user-database/src/index.ts:69]: Seems like this is unnecessary?

File: lambdas/db-provision-user-database/src/index.ts (hunk 1 above)
Comment (markdboyd) [reply lambdas/db-provision-user-database/src/index.ts:51]: An unrelated and possibly dumb question: do we need to be concerned about injection here when using `knex.raw`, since it's not escaping the input parameters?

File: lambdas/db-provision-user-database/variables.tf [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+  description = "Subnets to assign to the database provisioning lambda"
+}
+
+variable "rds_user_password" {
+  description = "Password to set for RDS db user"
+  type = string
+}
+
+variable "rds_access_secret_id" {
+  description = "AWS Secrets Manager secret ID containing jsonified db credentials containing at least host, password, port"
```
Comment (markdboyd) [inline lambdas/db-provision-user-database/variables.tf:27]: Would recommend the same edit suggested above

File: lambdas/db-provision-user-database/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  rootLoginSecret: string,
+  userLoginSecret: string,
+  prefix: string,
+  dbPassword: string,
+  engine: string,
+  dbClusterIdentifier: string,
+  env?: NodeJS.ProcessEnv,
+}
+
+export interface getSecretValue extends
```
Comment (markdboyd) [inline lambdas/db-provision-user-database/src/index.ts:14]: What is the reason for this?

File: example/data-persistence-tf/main.tf [hunk 16]
```diff
@@ -51,36 +8,31 @@ provider "aws" {
...
-  source = "../../tf-modules/data-persistence"
+resource "random_string" "db_pass" {
+  length  = 20
+  upper   = true
+  special = false
+}
 
-  prefix                     = var.prefix
-  subnet_ids                 = var.subnet_ids
+module "provision_database" {
```
Comment (markdboyd) [inline example/data-persistence-tf/main.tf:17]: So this Lambda invocation modle lives in data-persistence, because even though the cluster is created by a separate top level module to facilitate sharing a cluster across deployments, the database itself is unique to a deployment?

File: tf-modules/cumulus-rds-tf/main.tf [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+resource "aws_secretsmanager_secret" "rds_login" {
+  name_prefix = "cumulus_rds_db_cluster"
+}
+
+resource "aws_secretsmanager_secret_version" "rds_login" {
+  secret_id     = aws_secretsmanager_secret.rds_login.id
+  secret_string = jsonencode({
+    username            = var.db_username
```
Comment (markdboyd) [inline tf-modules/cumulus-rds-tf/main.tf:29]: I might be dense, but where do the initial `db_username` and `db_password` values come from? Oh, nevermind, looks like they are whatever you want and used below when the cluster is setup

File: tf-modules/cumulus-rds-tf/variables.tf [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+  default     = true
+}
+
+variable "cluster_identifier" {
+  description = "DB Itentifier for the RDS cluster that will be created"
+  type        = string
+  default     = "cumulus-rds-serverless-default-cluster"
+}
+
+variable "db_username" {
```
Comment (markdboyd) [inline tf-modules/cumulus-rds-tf/variables.tf:38]: I might consider renaming this to "admin_db_username". Same for the password. Just to make it clear how it is used. And update the descriptions accordingly

File: example/cumulus-tf/main.tf [hunk 19]
```diff
@@ -55,6 +57,9 @@ module "cumulus" {
   vpc_id            = var.vpc_id
   lambda_subnet_ids = var.subnet_ids
 
+  rds_security_group            = local.rds_security_group
```
Comment (markdboyd) [inline example/cumulus-tf/main.tf:60]: So the security group comes from the cluster module, then is input to data persistence module, then is read as output from the data persistence module. I guess it can't come from the cluster module directly without adding another remote state dependency here, which we don't want

File: lambdas/db-provision-user-database/variables.tf [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+  type = list(string)
+  description = "Subnets to assign to the database provisioning lambda"
+}
+
+variable "rds_user_password" {
+  description = "Password to set for RDS db user"
+  type = string
+}
+
+variable "rds_access_secret_id" {
```
Comment (markdboyd) [inline lambdas/db-provision-user-database/variables.tf:26]: So there are 2 different secrets managed throughout this code, right? One for admin access credentials and one for regular access credentials (to be used by Lamdas writing to RDS, etc). If so, I would recommend naming the one for admin access accordingly, so here `rds_admin_access_secret_id`. And maybe `rds_user_access_secret_id` for the other one?

File: tf-modules/cumulus-rds-tf/outputs.tf [hunk 21]
```diff
@@ -0,0 +1,16 @@
+output "security_group_id" {
+  value = aws_security_group.rds_cluster_access.id
+}
+
+output "rds_endpoint" {
+  value = aws_rds_cluster.core_team_cluster.endpoint
+}
+
+output "secret_arn" {
```
Comment (markdboyd) [inline tf-modules/cumulus-rds-tf/outputs.tf:9]:
```suggestion
output "admin_access_secret_arn" {
```

And same for the secret version below

File: lambdas/db-provision-user-database/README.md [hunk 22]
```diff
@@ -0,0 +1,18 @@
+# db-provision-user-database
+
+This module deploys a bootstrap lambda that, given a 'root' db user secret containing 'host', 'user', 'password' keys will:
+
+1) If configured database is not already created:
+  a) Add a database user based on the prefix and set the configured password
+  b) Add a database based on the passed in prefix and set the user to use that database
+2) If the configured database exists:
+  a) Update the user password to the configured value
+3) Write a database configuration object containing (at least) 'host', 'user', 'password' and 'database'
```
Comment (markdboyd) [inline lambdas/db-provision-user-database/README.md:10]:
```suggestion
3) Write an AWS Secrets Manager secret containing an object specify access credentials for acquiring a database connection (at least 'host', 'user', 'password' and 'database')
```

File: lambdas/db-provision-user-database/README.md [hunk 23]
```diff
@@ -0,0 +1,18 @@
+# db-provision-user-database
+
+This module deploys a bootstrap lambda that, given a 'root' db user secret containing 'host', 'user', 'password' keys will:
```
Comment (markdboyd) [inline lambdas/db-provision-user-database/README.md:3]:
```suggestion
This module deploys a bootstrap lambda that, given an AWS Secrets Manager secret specifying a 'root' database access credentials object containing 'host', 'user', 'password' keys, will:
```

File: lambdas/db-provision-user-database/README.md [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+  a) Update the user password to the configured value
+3) Write a database configuration object containing (at least) 'host', 'user', 'password' and 'database'
+
+## Input variables
+
+See [variables.tf](./variables.tf) for the input variables to this module and the default values for optional variables.
+
+## Outputs
+
+**database_credentials** - ARN of the AWS secrets manager user credentials secret
```
Comment (markdboyd) [inline lambdas/db-provision-user-database/README.md:18]:
```suggestion
**database_credentials_secret_arn** - ARN of the AWS secrets manager user credentials secret
```

Comment (markdboyd) [reply lambdas/db-provision-user-database/README.md:18]: I might opt for describing outputs in the "description" of each one rather than in a README, just to avoid drift

File: example/cumulus-tf/main.tf (hunk 19 above)
Comment (Jkovarik) [reply example/cumulus-tf/main.tf:60]: Right - given the Cluster module is potentially user supplied.     Also remember this is `cumulus-tf`, so this is for *us* - it's entirely possible for DAACs/Cirrus to bring in a remote state into their Core setup independently from data persistence.

File: example/data-persistence-tf/main.tf (hunk 16 above)
Comment (Jkovarik) [reply example/data-persistence-tf/main.tf:17]: Correct.    

This is analogous to how dynamo works - except AWS provisions Dynamo, and we create tables.   In postgres, they're just organized into "databases" within the database instance, and in theory also cross-organized by schema/schema perms, etc.    It made sense (to me) to handle 'database setup' here.  

The one thing this *does not* handle well is database deletion - but given that's a nearly trivial DBA task and this isn't the user module I chose not to address that here.

File: example/data-persistence-tf/outputs.tf [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+
+output "elasticsearch_alarms" {
+  value = module.data_persistence.elasticsearch_alarms
+}
+
+output "database_credentials_secret_arn" {
+  value = module.provision_database.database_credentials_secret_arn
+}
+
+output "rds_security_group" {
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply example/data-persistence-tf/outputs.tf:26]: Yes, correct.

File: example/data-persistence-tf/provision-database.tf (hunk 2 above)
Comment (Jkovarik) [reply example/data-persistence-tf/provision-database.tf:1]: No, not at all - it does nothing as this block is in the main.tf file.   Will remove.

Comment (Jkovarik) [reply example/data-persistence-tf/provision-database.tf]: 65f4ec063

File: example/data-persistence-tf/variables.tf (hunk 3 above)
Comment (Jkovarik) [inline example/data-persistence-tf/variables.tf:13]:
```suggestion
  description = "AWS Secrets Manager secret ID containing a JSON string of DB credentials (containing at least host, password, port as keys)"
```

Comment (Jkovarik) [reply example/data-persistence-tf/variables.tf:13]: @markdboyd  how about ^

File: example/rds-cluster-tf/variables.tf (hunk 7 above)
Comment (Jkovarik) [reply example/rds-cluster-tf/variables.tf:2]: The one used in the environment/.configuration file for CLI/and the AWS client library - I wrote the initial DB module that way for configurability and it bubbled up.   In theory this overrides the environment variable, if set.  

We actually don't do this elsewhere and rely *solely* on the environment variables, so that's inconsistent.  I can remove it.

File: lambdas/db-migration/main.tf (hunk 8 above)
Comment (Jkovarik) [reply lambdas/db-migration/main.tf:87]: So - 

This module assumes a VPC (is that valid now?   Other modules do not.... hmm)  - it in theory needs a security group for egress *if the database isn't in the VPC*.... but I expect the intention here was just that it needs a security group. 

So probably related to this work to remove that and just use the provided DB group if we're willing to concede that database work requires operations inside a VPC.   Or *really* if users want to use a non-VPC database setup they could pass in a SG that sets all egress or something. 

🤔

Comment (Jkovarik) [reply lambdas/db-migration/main.tf:87]: This raises another question re: if the networking settings in Core make sense around this for this work, given we're only using the DB security group if VPC is enabled and subnets are passed in.

File: lambdas/db-migration/src/index.ts (hunk 9 above)
Comment (Jkovarik) [reply lambdas/db-migration/src/index.ts:21]: Maybe.  I subconsciously hoping to avoid Core entanglements in this specific case, as for the moment the aws-client package doesn't really offer much here, but that's not really a good justification for not drying it up.

File: lambdas/db-migration/src/index.ts (hunk 10 above)
Comment (Jkovarik) [reply lambdas/db-migration/src/index.ts:27]: It's not so much it cannot be parsed in this line, as it didn't return a secret.    

I think we need to explicitly document the structure of this secret.... as for where.... 🤔  -- but *do* keep in mind that for the moment AWS SM if you create a secret via the console and tell it it's a postgres secret, it builds a form and JSONifies it for you - so there's some ... precedent for this format.      

In either case, if this isn't done right, it'll either throw a JSON.parse error, or the db connection will fail (and tell you as much) further on.

File: lambdas/db-migration/src/index.ts (hunk 11 above)
Comment (Jkovarik) [reply lambdas/db-migration/src/index.ts:38]: Hmm, I think moving it to the DB package is fine, so long as we leave connection timeouts and such as a param-configurable thing.    There's going to be configuration use cases for Core DB access and these scripts depending on the deployment setup.

File: lambdas/db-migration/src/index.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply lambdas/db-migration/src/index.ts:64]: Yes - the database startup time is not promised - we need to be defensive about read/write timeouts in a serverless-configured-to-go-fully-idle setup.

File: lambdas/db-provision-user-database/variables.tf (hunk 20 above)
Comment (Jkovarik) [reply lambdas/db-provision-user-database/variables.tf:26]: Hmm, yeah, that's a good suggestion, I'll see about sorting that up and down the PR

File: lambdas/db-migration/src/index.ts (hunk 10 above)
Comment (markdboyd) [reply lambdas/db-migration/src/index.ts:27]: Do we at least want to say that `No database credentials could be found for ${secret}`? So the error identifies the secret it attempted tor ead?

File: lambdas/db-migration/src/index.ts (hunk 9 above)
Comment (markdboyd) [reply lambdas/db-migration/src/index.ts:21]: The one thing about using the aws-client package is that it automatically switches to Localstack in unit test environments. That probably doesn't matter in this case, I assume.

In any case, I care less about using `aws-client`, but if we're going to copy `getSecretConnectionConfig` across multiple lambdas, it seems like it would be nice if it lived somewhere that it could be shared? Is there some extra complexity that creates that we don't want?

File: lambdas/db-migration/main.tf (hunk 8 above)
Comment (markdboyd) [reply lambdas/db-migration/main.tf:87]: Are you saying that the VPC/subnet settings for the DB migration should be required, not optional? It seems like it really depends if the use of the security group for RDS is optional? And if it isn't, why not?

Comment (Jkovarik) [reply lambdas/db-migration/main.tf:87]: No - I'm wondering if the SG for the database should be optional in Core.  

A VPC security group only works within a VPC - if you're running outside of one, I believe you need a DB security group with IP ranges, etc if you're trying to provide access to a RDS instance outside of a VPC.  

https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Overview.RDSSecurityGroups.html

tl;dr case 1) 

DB/lambda are in a VPC.   In this instance you can just put both an an accessor security group.  

2) DB or lambda not in a VPC    You then need a security group with rules specific to your access patterns, and the lambda has to be configured to allow egress to the <range the DB will deploy to - I'm a bit fuzzy on this> 


Case #1 makes obvious sense for most internal user use cases.      Case 2 might make sense if you were leveraging an existing postgres/rds database or some other more complicated network topology.

In thinking about this, the Core configuration concern is moot, as a lambda not in a VPC already has external access and can't manage DB access via a security group anyway.

File: example/db-migration-tf/variables.tf (hunk 4 above)
Comment (Jkovarik) [reply example/db-migration-tf/variables.tf:12]: Nothing, generally.   These are overrides for a situation where we know the data persistence layer isn't going to have that output (or the right output) for reason - in *theory* if the lookup has no key and this isn't set, it passes null to the submodule.    For `rds_access_secret_id` TF should error as `rds_access_secret_id` is required.   

This *might* be YAGNI in retrospect.

File: example/rds-cluster-tf/variables.tf (hunk 7 above)
Comment (Jkovarik) [reply example/rds-cluster-tf/variables.tf]: 7f3ef699b

File: lambdas/db-provision-user-database/src/index.ts (hunk 15 above)
Comment (Jkovarik) [reply lambdas/db-provision-user-database/src/index.ts]: Cruft from development only.  Removed: 46f6a70f8

File: lambdas/db-provision-user-database/src/index.ts (hunk 1 above)
Comment (Jkovarik) [reply lambdas/db-provision-user-database/src/index.ts:46]: Injection is always a serious concern.     To some degree this lambda is protected by IAM authentication/user setup, but there's always a possibility someone does something that allows this lambda to get run with a bad payload.   Nice catch. 

  I think we're going to need to iterate on this setup, regardless, but adding a fairly restrictive regexp sanitizer for now.

File: lambdas/db-provision-user-database/src/index.ts (hunk 13 above)
Comment (Jkovarik) [reply lambdas/db-provision-user-database/src/index.ts]: f12a2066b

File: lambdas/db-provision-user-database/src/index.ts (hunk 1 above)
Comment (Jkovarik) [reply lambdas/db-provision-user-database/src/index.ts:55]: f12a2066b

File: lambdas/db-provision-user-database/variables.tf (hunk 20 above)
Comment (Jkovarik) [reply lambdas/db-provision-user-database/variables.tf]: aa0dbf53d

File: tf-modules/cumulus-rds-tf/variables.tf (hunk 18 above)
Comment (Jkovarik) [reply tf-modules/cumulus-rds-tf/variables.tf]: 👍  on the descriptions.  That said this module's sole purpose is to setup the cluster, which makes the distinction less useful. 

2ddc06a65, d8f77d458

File: lambdas/db-migration/src/index.ts (hunk 10 above)
Comment (Jkovarik) [reply lambdas/db-migration/src/index.ts]: 3192e64a3 should probably cover both the key cases and do an up-front check

File: lambdas/db-provision-user-database/src/index.ts (hunk 1 above)
Comment (markdboyd) [reply lambdas/db-provision-user-database/src/index.ts:55]: Do we want to validate the database password as well?

File: lambdas/db-migration/main.tf (hunk 8 above)
Comment (markdboyd) [reply lambdas/db-migration/main.tf:87]: Yeah let's discuss this offline

File: lambdas/db-provision-user-database/variables.tf (hunk 14 above)
Comment (markdboyd) [reply lambdas/db-provision-user-database/variables.tf:27]: Seems like this description never got updated?

File: lambdas/db-provision-user-database/src/index.ts (hunk 1 above)
Comment (Jkovarik) [reply lambdas/db-provision-user-database/src/index.ts:57]: Yes indeed.  c61fb224f.     I think to some degree, the amount of concern greatly depends on if this winds up being a user deliverable - but given comments here, I'm going to add unit tests around it.

File: tf-modules/cumulus-rds-tf/outputs.tf (hunk 21 above)
Comment (Jkovarik) [reply tf-modules/cumulus-rds-tf/outputs.tf:9]: The output from this module is *implied* to be that though, as it's the entire purpose of the module is to setup the cluster/admin user.   Putting 'admin' on everything in this module seems redundant, if perhaps more explicit in integration. 

If you feel strongly about it, it's not a hill I want to die on either.

File: lambdas/db-provision-user-database/variables.tf (hunk 14 above)
Comment (Jkovarik) [reply lambdas/db-provision-user-database/variables.tf]: Indeed not, still working through comments, sorry. 

336b0727a

File: lambdas/db-provision-user-database/src/index.ts (hunk 1 above)
Comment (markdboyd) [reply lambdas/db-provision-user-database/src/index.ts:57]: Hate to keep being pedantic here, but forcing the password to only be alphanumeric is probably not great for security. Sorry.

File: tf-modules/cumulus-rds-tf/outputs.tf (hunk 21 above)
Comment (markdboyd) [reply tf-modules/cumulus-rds-tf/outputs.tf:9]: I guess it's not immediately obvious to me based on the module name or its purpose of setting up an RDS cluster that the `secret_arn` output is **necessarily** a secret containing admin access credentials. The `secret_arn` output from the module creating an RDS cluster could still be a secret containing almost anything, I think. Not a hill I'm going to die on either

File: lambdas/db-provision-user-database/src/index.ts (hunk 1 above)
Comment (Jkovarik) [reply lambdas/db-provision-user-database/src/index.ts:57]: Not entirely pedantic - to some degree the question is if a 60char alphanum versus a user putting in a 10char special char string here is meaningfully different for the purpose here.    If it weren't for IAM API restrictions as relates to scaling we wouldn't be *using* them at all.  

That said, it's not a huge issue to use a more permissive regexp, so long as we're careful about it.     I'll see if someone has implemented something more useful than my arbitrary regexp sanitization for PSQL.

Comment (Jkovarik) [reply lambdas/db-provision-user-database/src/index.ts:57]: Also if you don't like that, I'd expect you *really* don't like 

https://github.com/nasa/cumulus/pull/1826/files#diff-e7e41536cc3b3b76b76cdd76fefc9a23R11

Comment (Jkovarik) [reply lambdas/db-provision-user-database/src/index.ts]: Ok, I've added c14006ba6 (see comment about unit tests).   The real problem here is those queries cannot be parameterized... and really the right thing to do is make sure this lambda (and therefore access to the root db creds) are never run/accessed by anyone without deploy credentials (and therefore access to just change the db pw anyway).

File: tf-modules/cumulus-rds-tf/outputs.tf (hunk 21 above)
Comment (Jkovarik) [reply tf-modules/cumulus-rds-tf/outputs.tf]: Hmm, that's fair.  And "secret_arn" is pretty much SECRET THING.   Not terribly descriptive.  I see your point.

How about: 919095f70

File: lambdas/db-migration/main.tf (hunk 8 above)
Comment (Jkovarik) [reply lambdas/db-migration/main.tf:87]: @markdboyd  re-commenting so it gets on the agenda

File: lambdas/db-provision-user-database/src/index.ts (hunk 1 above)
Comment (markdboyd) [reply lambdas/db-provision-user-database/src/index.ts]: tl;dr - move this module to `example/` to reinforce the fact that it should not be released to users

File: lambdas/db-migration/main.tf (hunk 8 above)
Comment (markdboyd) [reply lambdas/db-migration/main.tf:87]: Discussed this and we're good. This code supports having an RDS cluster deployed inside and outside of a VPC
